observer/probers/prober.go

```diff
@@ -49,11 +56,17 @@ type Configurer interface {
 // the `MonConf`.
 type Settings map[string]interface{}
 
+// NormalizedKind normalizes the input string by stripping spaces and
+// transforming it into lowercase
+func NormalizedKind(kind string) string {
+	return strings.Trim(strings.ToLower(kind), " ")
+}
+
```

Comment:
Good idea here, however:

1. Ideally, we should keep this change (PR) small and only touch portions of the code that we need to.
1. I believe that `NormalizedKind` doesn't need to be capitalized as we're only calling it from within the package, not exporting it.

Reply:
Used in the observer package in `ObsConf.makeMonitors()` to ensure that `proberSpecificMetrics` doesn't end up with duplicate entries like "DNS" and "dns" or any permutation like that

Reply:
Ah, good catch! I see now that's derived directly from the parsed YAML. How about we make the current `Kind` field private (e.g. `kind`) and add a short accessor method to `MonConf`?

```go
func (c *MonConf) Kind() string {
	return strings.TrimSpace(strings.ToLower(c.kind))
}
```

1. You can then replace the line in `unmarshalConfigurer()` with `c.Kind()` 
1. Similarly in the `makeMonitors()` loop you can replace `m.Kind` with `m.Kind()`.
1. Finally, you can make this helper function private and modify it to return `strings.TrimSpace(strings.ToLower(kind))`

What I'm steering this towards is making sure that `kind` gets normalized on access so every part of our code deals with it consistently.

Reply:
Using a private field doesn't appear to work. I believe the reason is in the [docs](https://pkg.go.dev/gopkg.in/yaml.v2#Unmarshal) for `yaml.Unmarshal()`

> Struct fields are only unmarshalled if they are exported (have an upper case first letter) . . .

Now of course we could keep it a public field and just change its name, but then I fear we wouldn't really be solving the problem we set out to solve, in that the field would still totally be accessible from outside the package

Reply:
But I don't particularly see a way past that last part, so I think the only question is really what helper method(s)/function(s) do we provide (and implicitly hope all other code will use) and where?

Reply:
Though I suppose there's already some precedent for having a `Kind()` method attached to the object, as the `Prober` interface defines one.

(Sorry for the spam btw)

Reply:
Although I approve of having a helper function like this to ensure that GetConfigurer and Register always do the same thing, there's no need to export it and no need to call it from `makeMonitors`.

Instead, have that function simply set `kind := probers.GetConfigurer(m.Kind).Kind()`. Even if the return value from that function call isn't all-lowercase like a normalized one, you're still guaranteed that it will be the exact same value for all MonConfs of the same kind.

(Aside: this does make me question why Register takes a kind string as input, when it could just get that from the .Kind() method on its second Configurer argument...)

Reply:
So there might be some confusion there. There are `Configurer`s and there are `Prober`s, and currently the `Prober`s (constructed at configuration time, not on import) are the only ones that have the `Kind()` method. We could do it the way you suggest, but that does mean we'd need to add the `Kind()` method to the `Configurer` interface and go implement those everywhere. I believe this also answers the question you pose in your aside

Reply:
Ah yes you're right, I thought that `.Kind()` was already on the `Configurer` interface.

And honestly I think that's the cleanest solution. We can clean clean up the seeming duplication of work between `prober.Kind()` and `conf.Kind()` in a future refactoring that essentially gets rid of the configurer types entirely.

Reply:
So I've run into a snag here and would like input before proceeding. `MockConfigurer` constructs its prober by passing all of the yaml values into it, including its kind. `MockProber.Kind()` simply returns that string as-is, meaning its value depends on what was specified in the config. However, `MockConfigurer.Kind()` wouldn't be able to do this, since the config hasn't been read at the time of construction; it has to return a constant. The code is simple enough, just `return "Mock"`. But it breaks what could be an implicit assumption here that `Configurer.Kind()` will always equal `Prober.Kind()`. So my question is: are we okay with prober kinds not always equaling their configurer's kind, or should we maybe change `MockProber` to return a constant and declare that this isn't allowed?

Reply:
Yep, it's a nasty tangle, and it's what I hope to clean up better in the future. I think, for now, it's fine for the Configurer's Kind to always be hard-coded and the Prober's Kind to be whatever was in the yaml. After all, as long as the Configurer's Kind is what was used as the map key in the Registry, we're guaranteed that the Configurer and Prober will have Kinds that are identical modulo capitalization, because otherwise we wouldn't have been able to look up the correct configurer for this prober.

Reply:
Done!